src/quote/swap/adapters/zeroex.ts

```diff
@@ -0,0 +1,165 @@
+import axios, { AxiosRequestHeaders } from 'axios'
+
+import { ChainId } from 'constants/chains'
+import { SwapQuoteProvider, SwapQuoteRequest } from 'quote/swap/interfaces'
+
+type ZeroExApiSwapRequest = {
+  buyAmount?: string
+  buyToken: string
+  sellAmount?: string
+  sellToken: string
+  // 0xAPI expects percentage as value between 0-1 e.g. 5% -> 0.05
+  slippagePercentage?: string
+}
+
+export type ZeroExApiSwapResponseOrder = {
+  source: string
+  fillData: {
+    path?: string
+    pool?: {
+      tokens: string[]
+      poolAddress: string
+    }
+    tokenAddressPath?: string[]
+  }
+}
+
+export type ZeroExApiSwapResponseOrderBalancer = {
+  source: string
+  fillData: {
+    assets: string[]
+    chainId: number
+    swapSteps: {
+      poolId: string
+    }[]
+    vault: string
+  }
+}
+
+export type ZeroExApiSwapResponseOrderSushi = {
+  source: string
+  makerToken: string
+  takerToken: string
+  makerAmount: string
+  takerAmount: string
+  fillData: {
+    tokenAddressPath: string[]
+    router: string
+  }
+  fill: {
+    input: string
+    output: string
+    adjustedOutput: string
+    gas: number
+  }
+}
+
+export type ZeroExApiSwapResponse = {
+  buyAmount: string
+  buyTokenAddress: string
+  data: string
+  orders?:
+    | ZeroExApiSwapResponseOrder[]
+    | ZeroExApiSwapResponseOrderBalancer[]
+    | ZeroExApiSwapResponseOrderSushi[]
+  sellAmount: string
+  sellTokenAddress: string
+}
+
+export class ZeroExSwapQuoteProvider implements SwapQuoteProvider {
+  /**
+   * @param baseUrl              The base url (default: https://api.0x.org, watch rate limits)
+   * @param affiliateAddress    (Optional) Affiliate address
+   * @param headersOverride      (Optional) Override for headers
+   * @param swapPathOverride     (Optional) Override of the API path - in case your using a custom path format e.g. through a proxy
+   */
+  constructor(
+    private readonly baseUrl: string | null = null,
+    private readonly affiliateAddress: string | null = null,
+    private readonly headersOverride: AxiosRequestHeaders | null = null,
+    private readonly swapPathOverride: string | null = null
+  ) {}
+
+  /**
+   * Builds the 0x API URL.
+   * @param path     An API path in the form of /swap/v1/quote
+   * @param query    A query constructed with URLSearchParams
+   * @param chainId  ID of the network
+   */
+  public buildUrl(path: string, query: string, chainId: number): string {
+    const baseUrl = this.getBaseUrl(chainId)
+    let url = `${baseUrl}${path}?${query}`
+    if (this.affiliateAddress) {
+      url += `&affiliateAddress=${this.affiliateAddress}`
+    }
+    return url
+  }
+
+  /**
+   * Get a swap quote as described in:
+   * https://0x.org/docs/0x-swap-api/api-references/get-swap-v1-quote
+   *
+   * @param request An instance of type SwapQuoteRequest
+   */
+  public async getSwapQuote(
+    request: SwapQuoteRequest
+  ): Promise<ZeroExApiSwapResponse | null> {
```

Comment:
I'd still need to change this to the protocol's return type: `SwapQuote`. Just wanted to test first if this would all work.